repo/fsrepo/migrations/httpfetcher.go

```diff
@@ -95,9 +108,63 @@ func (f *HttpFetcher) Fetch(ctx context.Context, filePath string) ([]byte, error
 	}
 	defer rc.Close()
 
-	return io.ReadAll(rc)
+	return carStreamToFileBytes(ctx, rc)
 }
 
 func (f *HttpFetcher) Close() error {
 	return nil
 }
+
+func carStreamToFileBytes(ctx context.Context, r io.Reader) ([]byte, error) {
+	// Create temporary block datastore and dag service.
+	db := dssync.MutexWrap(ds.NewMapDatastore())
+	bs := bstore.NewBlockstore(db)
+	ds := dag.NewDAGService(bservice.New(bs, offline.Exchange(bs)))
+
+	defer ds.Blocks.Close()
+	defer db.Close()
+
+	// Create CAR reader
+	car, err := gocarv2.NewBlockReader(r)
+	if err != nil {
+		return nil, fmt.Errorf("error creating car reader: %s", err)
+	}
+
+	// Add all blocks to the blockstore.
+	for {
+		block, err := car.Next()
+		if err != nil && err != io.EOF {
+			return nil, fmt.Errorf("error reading block from car: %s", err)
+		} else if block == nil {
+			break
+		}
+
+		err = bs.Put(ctx, block)
+		if err != nil {
+			return nil, fmt.Errorf("error putting block in blockstore: %s", err)
+		}
+	}
+
+	if len(car.Roots) != 1 {
+		return nil, errors.New("multiple-root CAR unexpected")
+	}
+
+	// Get node from DAG service with the file.
+	node, err := ds.Get(ctx, car.Roots[0])
```

Comment:
This is incorrect, basically you download a car file in order to hash the data locally however the root hash you are using is provided by the remote peer.
An attacker could give you a malicious root that include malware and you would be oblivious to it.
(it's fine because the attacker is the gateway which default to us, but this change would basically do nothing to the trust model, we want to stop trusting the gateway :slightly_smiling_face:)

You need to use the `path.Join(f.distPath, filePath)` above and perform the unixfs path walk manually, I can do that with feather if you want (I think this is gonna be easier.)

Reply:
> You need to use the `path.Join(f.distPath, filePath)` above and perform the unixfs path walk manually, I can do that with feather if you want (I think this is gonna be easier.)

Actually it might be possible to reuse handler code `ipfs get` with the in memory datastore.
Which wouldn't be that hard to do (it would be extremely ugly tho).

Reply:
@Jorropo if you're willing to, feel free to make any changes to this PR. I completely forgot about it!